rig/place_and_route/constraints.py

```diff
@@ -0,0 +1,125 @@
+implementation-specific constraints seperately.
+"""
+
+
+class LocationConstraint(object):
+    """Unconditionally place a vertex on a specific chip.
+
+    Attributes
+    ----------
+    vertex : object
+        The user-supplied object representing the vertex.
+    location : (x, y)
+        The x- and y-coordinates of the chip the vertex must be placed on.
+    """
+
+    __slots__ = ["vertex", "location"]
+
+    def __init__(self, vertex, location):
+        self.vertex = vertex
+        self.location = location
+
+
+class ReserveResourceConstraint(object):
+    """Reserve a range of a resource on all or a specific chip.
+
+    For example, this can be used to reserve areas of SDRAM used by the system
+    software to prevent allocations occurring there.
+
+    Note: Reserved ranges must *not* be be partly or fully outside the
+    available resources for a chip nor may they overlap with one another.
+    Violation of these rules will result in undefined behaviour.
+
+    Note: placers are obliged by this constraint to subtract the reserved
+    resource from the total available resource but *not* to determine whether
+    the remaining resources include sufficient continuous ranges of resource
+    for their placement. Users should thus be extremely careful reserving
+    resources which are not immediately at the beginning or end of a resource
+    range.
+
+    Attributes
+    ----------
+    resource : object
+        A resource identifier for the resource being reserved.
+    reservation : :py:class:`slice`
+        The range over that resource which must not be used.
+    location : (x, y) or None
+        The chip to which this reservation applies. If None then the
+        reservation applies globally.
+    """
+
+    __slots__ = ["resource", "reservation", "location"]
+
+    def __init__(self, resource, reservation, location=None):
+        self.resource = resource
+        self.reservation = reservation
+        self.location = location
+
+
+class AlignResourceConstraint(object):
+    """Force alignment of start-indices of resource ranges.
+
+    For example, this can be used to ensure assignments into SDRAM are word
+    aligned.
+
+    Note: placers are not obliged to be aware of or compensate for wastage of a
+    resource due to this constraint and so may produce impossible placements
+    when in the even of large numbers of individual items using a non-aligned
+    width block of resource.
+
+    Attributes
+    ----------
+    resource : object
+        A resource identifier for the resource to align.
+    alignment : int
+        The number of which all assigned start-indices must be a multiple.
+    """
+
+    __slots__ = ["resource", "alignment"]
+
+    def __init__(self, resource, alignment):
+        self.resource = resource
+        self.alignment = alignment
+
+
+class RouteEndpointConstraint(object):
+    """Force the endpoint of a path through the network to be a particular
+    route.
```

Comment:
Good catch, this one is a pain in the arse since it originally was called a Link constraint (and was on links) but it would be legitimately useful to constrain to other endpoints too, e.g. the monitor core.

Some minor re-wording will occur which may/may not improve the situation slightly.